File: LiberatePDF2-Rest-Service/src/main/java/rocks/huwi/liberatepdf2/restservice/restrictionsremover/RestrictionsRemoverService.java
package rocks.huwi.liberatepdf2.restservice.restrictionsremover;

import org.springframework.scheduling.annotation.Async;

import rocks.huwi.liberatepdf2.restservice.Pdf;

/**
 * Removes restrictions from files.
 */
public interface RestrictionsRemoverService {
	/**
	 * Remove restrictions from file.
	 *
	 * @param original
	 *            file to remove restrictions from
	 * @return
	 */
	public void removeRestrictions(Pdf pdf);

	/**
	 * Enqueue a task to remove removes restrictions
	 *
	 * @param original
	 *            file to remove restrictions from
	 * @return
	 */
	@Async
	public void removeRestrictionsAsync(Pdf pdf);
	
	/**
	 * Gets the count of processed items.
	 * @return
	 */
	public Long getItemsCount();
	
	/**
	 * Gets the count of unsuccessfully processed items.
	 * @return
	 */
	public Long getFailedItemsCount();
}

File: LiberatePDF2-Rest-Service/src/main/java/rocks/huwi/liberatepdf2/restservice/restrictionsremover/RestrictionsRemoverService.kt
package rocks.huwi.liberatepdf2.restservice.restrictionsremover

import org.springframework.scheduling.annotation.Async
import rocks.huwi.liberatepdf2.restservice.Pdf

/**
 * Removes restrictions from files.
 */
interface RestrictionsRemoverService {
    /**
     * Gets the count of unsuccessfully processed items.
     *
     * @return
     */
    val failedItemsCount: Long

    /**
     * Gets the count of processed items.
     *
     * @return
     */
    val itemsCount: Long

    /**
     * Remove restrictions from file.
     *
     * @param original
     * file to remove restrictions from
     * @return
     */
    fun removeRestrictions(pdf: Pdf)

    /**
     * Enqueue a task to remove restrictions
     *
     * @param original
     * file to remove restrictions from
     * @return
     */
    @Async
    fun removeRestrictionsAsync(pdf: Pdf)
}